calico_containers/tests/unit/diags_test.py

```diff
@@ -0,0 +1,196 @@
+        """
+        Test save_diags for calicoctl diags command
+        """
+        # Set up mock objects
+        m_tempfile.mkdtemp.return_value = '/temp/dir'
+        date_today = '2015-7-24_09_05_00'
+        m_datetime.strftime.return_value = date_today
+        m_socket.return_value = 'hostname'
+        m_sh_command_return = Mock(autospec=True)
+        m_sh_command.return_value = m_sh_command_return
+        m_datetime.today.return_value = 'diags-07242015_090500.tar.gz'
+        m_os_mkdir.return_value = True
+        # The DatastoreClient contains an etcd Client
+        # The etcd Client reads in a list of children of type EtcdResult
+        # The children are accessed by calling get_subtree method on the etcd Client
+        m_datastore_client = Mock(spec=DatastoreClient)
+        m_datastore_client.etcd_client = Mock(spec=Client)
+        m_datastore_data = Mock(spec=EtcdResult)
+        m_child_1 = EtcdResult(node={'dir': True, 'key': 666})
+        m_child_2 = EtcdResult(node={'key': 555, 'value': 999})
+        m_datastore_data.get_subtree.return_value = [m_child_1, m_child_2]
+        m_datastore_client.etcd_client.read.return_value = m_datastore_data
+        m_DatastoreClient.return_value = m_datastore_client
+
+        # Set up arguments
+        log_dir = '/log/dir'
+        temp_dir = '/temp/dir/'
+        diags_dir = temp_dir + 'diagnostics'
+
+        # Call method under test
+        diags.save_diags(log_dir, upload=True)
+
+        # Assert
+        m_subprocess.call.assert_called_once_with(
+            ["docker", "exec", "calico-node", "pkill", "-SIGUSR1", "felix"])
+        m_tempfile.mkdtemp.assert_called_once_with()
+        m_os_mkdir.assert_called_once_with(diags_dir)
+        m_open.assert_has_calls([
+            call(diags_dir + '/date', 'w'),
+            call().__enter__().write('DATE=%s' % date_today),
+            call(diags_dir + '/hostname', 'w'),
+            call().__enter__().write('hostname'),
+            call(diags_dir + '/netstat', 'w'),
+            call().__enter__().writelines(m_sh_command_return()),
+            call(diags_dir + '/route', 'w'),
+            call().__enter__().write('route --numeric\n'),
+            call().__enter__().writelines(m_sh_command_return()),
+            call().__enter__().write('ip route\n'),
+            call().__enter__().writelines(m_sh_command_return()),
+            call().__enter__().write('ip -6 route\n'),
+            call().__enter__().writelines(m_sh_command_return()),
+            call(diags_dir + '/iptables', 'w'),
+            call().__enter__().writelines(m_sh_command_return()),
+            call(diags_dir + '/ipset', 'w'),
+            call().__enter__().writelines(m_sh_command_return()),
+            call(diags_dir + '/etcd_calico', 'w'),
+            call().__enter__().write('dir?, key, value\n'),
+            call().__enter__().write('DIR,  666,\n'),
+            call().__enter__().write('FILE, 555, 999\n')
+        ], any_order=True)
```

Comment:
Seems like the save_diags function is pretty large.  Should we have some helper functions instead that can be called to slim down the function for things like writing netstat, writing routes, writing iptables, etc...